### ISSUE TITLE
The optimal image aspect ratio for single-image calibration.

### ISSUE THREAD

**Issue by ChenYutongTHU** [facebookresearch/map-anything#8]
Dear authors,

Thanks for this great work!

I have a question about the single-image calibration task.

In my experiment, I found that center-cropping the image led to the change of output focal lengths.

For example, for the same image from ETH3D dataset, I cropped the top and bottom pixels with a margin of 14, 2x14, 3x14, 4x14, and fed the cropped image into the model. The predicted focal length will be different for different heights. The variation is non-trivial and the FoV angle error can sometimes be as large as 4 deg.

```python
import os
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"

# Required imports
import torch
from mapanything.models import MapAnything
from mapanything.utils.image import load_images

# Get inference device
device = "cuda" if torch.cuda.is_available() else "cpu"

# Init model - This requries internet access or the huggingface hub cache to be pre-downloaded
# For Apache 2.0 license model, use "facebook/map-anything-apache"
model = MapAnything.from_pretrained("facebook/map-anything").to(device)


'''
1 PINHOLE 6180 4130 3409.58 3409.44 3090.16 2064.73
courtyard/dslr_calibration_undistorted/example_images/eth3d_mono/DSC_0286.JPG
'''

images = "./eth3d_mono"  # or ["path/to/img1.jpg", "path/to/img2.jpg", ...]
views0 = load_images(images)
gt_fx_native, gt_fy_native = torch.tensor(3409.58), torch.tensor(3409.44)
scale_factor = views0[0]['img'].shape[-1]/6180
gt_fx, gt_fy = gt_fx_native*scale_factor, gt_fy_native*scale_factor

for i in range(0,5):
    if i==0:
        views = views0
    else:
        # crop the top and bottom 14*i pixels, this will not change the focal length.
        views = [{
            'img':v['img'][:,:,14*i:-1*14*i,:],
            'true_shape':v['img'][:,:,14*i:-1*14*i,:].shape[2:],
            'idx':v['idx'],
            'instance':v['instance'],
            'data_norm_type':['dinov2']
        } for v in views0]
    predictions = model.infer(
        views,                            # Input views
        memory_efficient_inference=False, # Trades off speed for more views (up to 2000 views on 140 GB)
        use_amp=True,                     # Use mixed precision inference (recommended)
        amp_dtype="bf16",                 # bf16 inference (recommended; falls back to fp16 if bf16 not supported)
        apply_mask=True,                  # Apply masking to dense geometry outputs
        mask_edges=True,                  # Remove edge artifacts by using normals and depth
        apply_confidence_mask=False,      # Filter low-confidence regions
        confidence_percentile=10,         # Remove bottom 10 percentile confidence pixels
    )
    h, w = predictions[0]["conf"].shape[-2:]
    print(f'Crop {14*i} pixels from top and bottom, image size: {w}x{h}')
    gt_fovx = torch.atan2(torch.tensor(1/2*w), gt_fx) *2 /torch.pi*180
    gt_fovy = torch.atan2(torch.tensor(1/2*h), gt_fy) *2 /torch.pi*180
    pred_fx, pred_fy = predictions[0]["intrinsics"][0,0,0].cpu(), predictions[0]["intrinsics"][0,1,1].cpu()
    pred_fovx = torch.atan2(torch.tensor(1/2*w), pred_fx) *2 /torch.pi*180
    pred_fovy = torch.atan2(torch.tensor(1/2*h), pred_fy) *2 /torch.pi*180
    print(f'pred_fx: {pred_fx:.2f}, gt_fx: {gt_fx}')
    print(f'pred_fovx: {pred_fovx:.2f}, gt_fovx: {gt_fovx:.2f}, diff: {(pred_fovx-gt_fovx).abs():.2f} (deg)' )
    print(f'pred_fy: {pred_fy:.2f}, gt_fy: {gt_fy}')
    print(f'pred_fovy: {pred_fovy:.2f}, gt_fovy: {gt_fovy:.2f}, diff: {(pred_fovy-gt_fovy).abs():.2f} (deg)' )
    print('-----')
```

The result is 

```
Crop 0 pixels from top and bottom, image size: 518x336
pred_fx: 276.01, gt_fx: 285.78680419921875
pred_fovx: 86.36, gt_fovx: 84.37, diff: 1.99 (deg)
pred_fy: 274.90, gt_fy: 285.7750549316406
pred_fovy: 62.86, gt_fovy: 60.90, diff: 1.96 (deg)
-----
Crop 14 pixels from top and bottom, image size: 518x308
pred_fx: 273.61, gt_fx: 285.78680419921875
pred_fovx: 86.86, gt_fovx: 84.37, diff: 2.49 (deg)
pred_fy: 263.18, gt_fy: 285.7750549316406
pred_fovy: 60.67, gt_fovy: 56.64, diff: 4.03 (deg)
-----
Crop 28 pixels from top and bottom, image size: 518x280
pred_fx: 272.05, gt_fx: 285.78680419921875
pred_fovx: 87.19, gt_fovx: 84.37, diff: 2.81 (deg)
pred_fy: 270.72, gt_fy: 285.7750549316406
pred_fovy: 54.69, gt_fovy: 52.20, diff: 2.49 (deg)
-----
Crop 42 pixels from top and bottom, image size: 518x252
pred_fx: 276.00, gt_fx: 285.78680419921875
pred_fovx: 86.36, gt_fovx: 84.37, diff: 1.99 (deg)
pred_fy: 276.48, gt_fy: 285.7750549316406
pred_fovy: 49.00, gt_fovy: 47.59, diff: 1.41 (deg)
-----
Crop 56 pixels from top and bottom, image size: 518x224
pred_fx: 287.41, gt_fx: 285.78680419921875
pred_fovx: 84.05, gt_fovx: 84.37, diff: 0.32 (deg)
pred_fy: 262.49, gt_fy: 285.7750549316406
pred_fovy: 46.21, gt_fovy: 42.80, diff: 3.41 (deg)
-----
```

I wonder how I should set the image aspect ratio to get the best result of the model? 

Great thanks!

**Comment by ChenYutongTHU** [comment]
![Image](https://github.com/user-attachments/assets/f5658c0e-88bc-4e95-8f35-3cca05dc169d)

**Comment by Nik-V9** [comment]
Hi, Thanks for testing this out!

When we train MapAnything, we use a fixed set of aspect ratios and resolutions and find that performance degrades when the inputs deviate from this.

Hence, when we load images, we take care to convert them to a fixed set of pre-defined resolution mappings: https://github.com/facebookresearch/map-anything/blob/efd47b58d29f4d7114a2a0d2a4c1361cb9815602/mapanything/utils/image.py#L39

In the future, similar to MoGe-2 & Pi3, we aim to use a more continuous interpolation of aspect ratio and resolutions during training, to support any resolution input.

For now, I would recommend using an aspect ratio and resolution which matches the mapping in the image utility folder.

**Comment by ChenYutongTHU** [comment]
Hi. Thanks for the prompt reply! This is quite interesting.
I have one more question regarding the multi-view depth evaluation.

Before computing the rmse and inlier-ratio, the gt point maps and prediction point maps are normalized independently, each divided by their own averages. Do I understand it correctly?

https://github.com/facebookresearch/map-anything/blob/efd47b58d29f4d7114a2a0d2a4c1361cb9815602/benchmarking/dense_n_view/benchmark.py#L173-L187

Other works such as VGGT and FAST3R choose to estimate the Sim3D alignment between two point clouds via Umeyama, before computing the 3D error.  

Could you help explain the reason for this evaluation design?  Do you also have evaluation results with Umeyama alignment?
Thank you so much!

**Comment by ChenYutongTHU** [comment]
Going back to the problem of fixed ratio.
When I set the ETH3D images to 518x294, which is included in 
https://github.com/facebookresearch/map-anything/blob/efd47b58d29f4d7114a2a0d2a4c1361cb9815602/mapanything/utils/image.py#L40-L52

the network cannot produce correct focal length and the angular error is around 4 deg.

I guess it  is not sufficient to just pick one of the fixed ratios, and we need to try out different aspect ratios for different datasets/domains?

**Comment by Nik-V9** [comment]
> Before computing the rmse and inlier-ratio, the gt point maps and prediction point maps are normalized independently, each divided by their own averages. Do I understand it correctly?

Yes, where the key thing is that the gt pointmap validity mask is used across both the gt and predictions. This ensures that the same set of pixels/3D points are used for normalization across both ground truth and prediction.

> Other works such as VGGT and FAST3R choose to estimate the Sim3D alignment between two point clouds via Umeyama, before computing the 3D error.
> 
> Could you help explain the reason for this evaluation design? Do you also have evaluation results with Umeyama alignment? 

I believe that both would result in similar trends. The motivation for doing a umeyama alignment or a normalization like what we do (which is similar to what *3R codebases do for normalization in the loss function) is the same where you want to find an ideal scale factor which abstracts away the up to scale nature of predictions from VGGT and other non-metric models. However, one drawback with Umeyama alignment is that you can also do a scale shift alignment which is not the exact capability you want to measure. In theory, the reconstructions should only be up to scale, shift should not be a unknown or wrong. Hence why we choose to go with the normalization route for benchmarking as opposed to doing some optimization which minimizes alignment error (can also be very error prone and can add systematic bias to benchmarking: https://arxiv.org/abs/2203.08122).

**Comment by Nik-V9** [comment]
> Going back to the problem of fixed ratio. When I set the ETH3D images to 518x294, which is included in

Can you provide more details on how you are setting it to 518 x 294?

**Comment by ChenYutongTHU** [comment]
> > Before computing the rmse and inlier-ratio, the gt point maps and prediction point maps are normalized independently, each divided by their own averages. Do I understand it correctly?
> 
> Yes, where the key thing is that the gt pointmap validity mask is used across both the gt and predictions. This ensures that the same set of pixels/3D points are used for normalization across both ground truth and prediction.
> 
> > Other works such as VGGT and FAST3R choose to estimate the Sim3D alignment between two point clouds via Umeyama, before computing the 3D error.
> > Could you help explain the reason for this evaluation design? Do you also have evaluation results with Umeyama alignment?
> 
> I believe that both would result in similar trends. The motivation for doing a umeyama alignment or a normalization like what we do (which is similar to what *3R codebases do for normalization in the loss function) is the same where you want to find an ideal scale factor which abstracts away the up to scale nature of predictions from VGGT and other non-metric models. However, one drawback with Umeyama alignment is that you can also do a scale shift alignment which is not the exact capability you want to measure. In theory, the reconstructions should only be up to scale, shift should not be a unknown or wrong. Hence why we choose to go with the normalization route for benchmarking as opposed to doing some optimization which minimizes alignment error (can also be very error prone and can add systematic bias to benchmarking: https://arxiv.org/abs/2203.08122).

Thanks for your detailed reply!  That makes some sense.

**Comment by ChenYutongTHU** [comment]
> Can you provide more details on how you are setting it to 518 x 294?

I used the same code above, only changing the cropping to 21 pixels for the top and bottom

**Comment by Nik-V9** [comment]
Closing issue, please reopen if further assistance is needed.
